nursery_store.go

```diff
@@ -0,0 +1,1377 @@
+	// this channel point.
+	return hghtBucket.CreateBucketIfNotExists(chanBytes)
+}
+
+// getHeightChanBucket retrieves an existing height-channel bucket from the
+// nursery store, using the provided block height and channel point. if the
+// bucket does not exist, or any bucket along its path does not exist, a nil
+// value is returned.
+func (ns *nurseryStore) getHeightChanBucket(tx *bolt.Tx,
+	height uint32, chanPoint *wire.OutPoint) *bolt.Bucket {
+
+	// Retrieve the existing height bucket from this nursery store.
+	hghtBucket := ns.getHeightBucket(tx, height)
+	if hghtBucket == nil {
+		return nil
+	}
+
+	// Serialize the provided channel point, which generates the key for
+	// looking up the proper height-channel bucket inside the height bucket.
+	var chanBuffer bytes.Buffer
+	if err := writeOutpoint(&chanBuffer, chanPoint); err != nil {
+		return nil
+	}
+	chanBytes := chanBuffer.Bytes()
+
+	// Finally, return the height bucket specified by the serialized channel
+	// point.
+	return hghtBucket.Bucket(chanBytes)
+}
+
+// forEachHeightPrefix enumerates all outputs at the given height whose state
+// prefix matches that which is provided. This is used as a subroutine to help
+// enumerate crib and kindergarten outputs at a particular height. The callback
+// is invoked with serialized bytes retrieved for each output of interest,
+// allowing the caller to deserialize them into the appropriate type.
+func (ns *nurseryStore) forEachHeightPrefix(tx *bolt.Tx, prefix []byte,
+	height uint32, callback func([]byte) error) error {
+
+	// Start by retrieving the height bucket corresponding to the provided
+	// block height.
+	hghtBucket := ns.getHeightBucket(tx, height)
+	if hghtBucket == nil {
+		return nil
+	}
+
+	// Using the height bucket as a starting point, we will traverse its
+	// entire two-tier directory structure, and filter for outputs that have
+	// the provided prefix. The first layer of the height bucket contains
+	// buckets identified by a channel point, thus we first create list of
+	// channels contained in this height bucket.
+	var channelsAtHeight [][]byte
+	if err := hghtBucket.ForEach(func(chanBytes, _ []byte) error {
+		channelsAtHeight = append(channelsAtHeight, chanBytes)
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	// As we enumerate the outputs referenced in this height bucket, we will
+	// need to load the serialized value of each output, which is ultimately
+	// stored its respective channel bucket. To do so, we first load the
+	// top-level chain bucket, which should already be created if we are at
+	// this point.
+	chainBucket := tx.Bucket(ns.pfxChainKey)
+	if chainBucket == nil {
+		return nil
+	}
+
+	// Additionally, grab the chain index, which we will facilitate queries
+	// for each of the channel buckets of each of the channels in the list
+	// we assembled above.
+	chanIndex := chainBucket.Bucket(channelIndexKey)
+	if chanIndex == nil {
```

Comment:
Should also probably be an error.

Reply:
got to poke around on these comments last night, all three should be returning errors. in doing so, it exposed some bugs in the pruning logic, and ultimately forced me to tighten that up a good bit ⚡️